File: ui-automation/selenium-javascript/pages/EcommercePage.js
const { By, until } = require('selenium-webdriver');

class EcommercePage {
    constructor(driver) {
        this.driver = driver;
        
        // Page element locators
        this.searchInput = By.name('search');
        this.homeButton = By.xpath("//span[@class='title' and contains(text(), ' Home')]");
        this.specialButton = By.xpath("//div[@id='widget-navbar-217834']//li[contains(text(), ' Special')]");
        this.blogButton = By.xpath("//div[@id='widget-navbar-217834']//li[contains(text(), ' Blog')]");
        this.megaMenuButton = By.xpath("//div[@id='widget-navbar-217834']//li[contains(text(), ' Mega Menu')]");
        this.specialOffersHeader = By.xpath("//h1[@class='h3' and contains(text(), 'Special Offers')]");
        this.latestArticlesHeader = By.xpath("//h3[contains(text(), 'Latest Articles')]");
        this.sonyVaioImage = By.xpath("//img[@alt='Sony VAIO']");
        this.productThumb = By.className('product-thumb');
        this.productLayout = By.className('product-layout');
        this.noResultsMessage = By.xpath("//div[contains(text(), 'There is no product that matches the search criteria.')]");
        this.showingResultsText = By.xpath("//div[@class='text-right' and contains(text(), 'Showing')]");
    }

    /**
     * Navigate to homepage
     */
    async navigateToHomepage() {
        await this.driver.get('https://ecommerce-playground.lambdatest.io/');
    }

    /**
     * Click home button
     */
    async clickHomeButton() {
        const element = await this.driver.wait(until.elementLocated(this.homeButton), 10000);
        await this.driver.wait(until.elementIsEnabled(element), 10000);
        await element.click();
    }

    /**
     * Click special button
     */
    async clickSpecialButton() {
        const element = await this.driver.wait(until.elementLocated(this.specialButton), 10000);
        await this.driver.wait(until.elementIsEnabled(element), 10000);
        await element.click();
    }

    /**
     * Click blog button
     */
    async clickBlogButton() {
        const element = await this.driver.wait(until.elementLocated(this.blogButton), 10000);
        await this.driver.wait(until.elementIsEnabled(element), 10000);
        await element.click();
    }

    /**
     * Click mega menu button
     */
    async clickMegaMenuButton() {
        const element = await this.driver.wait(until.elementLocated(this.megaMenuButton), 10000);
        await this.driver.wait(until.elementIsEnabled(element), 10000);
        await element.click();
    }

    /**
     * Perform search
     */
    async performSearch(searchTerm) {
        const searchElement = await this.driver.wait(until.elementLocated(this.searchInput), 10000);
        await searchElement.click();
        await searchElement.clear();
        await searchElement.sendKeys(searchTerm);
        await searchElement.submit();
    }

    /**
     * Check if search input is visible
     */
    async isSearchInputVisible() {
        try {
            const element = await this.driver.wait(until.elementLocated(this.searchInput), 10000);
            return await element.isDisplayed();
        } catch (error) {
            return false;
        }
    }

    /**
     * Check if home button is visible
     */
    async isHomeButtonVisible() {
        try {
            const element = await this.driver.findElement(this.homeButton);
            return await element.isDisplayed();
        } catch (error) {
            return false;
        }
    }

    /**
     * Check if special offers header is visible
     */
    async isSpecialOffersHeaderVisible() {
        try {
            const element = await this.driver.wait(until.elementLocated(this.specialOffersHeader), 10000);
            return await element.isDisplayed();
        } catch (error) {
            return false;
        }
    }

    /**
     * Check if latest articles header is visible
     */
    async isLatestArticlesHeaderVisible() {
        try {
            const element = await this.driver.wait(until.elementLocated(this.latestArticlesHeader), 10000);
            return await element.isDisplayed();
        } catch (error) {
            return false;
        }
    }

    /**
     * Check if Sony VAIO image is visible
     */
    async isSonyVaioImageVisible() {
        try {
            const element = await this.driver.wait(until.elementLocated(this.sonyVaioImage), 10000);
            return await element.isDisplayed();
        } catch (error) {
            return false;
        }
    }

    /**
     * Check if products are visible
     */
    async areProductsVisible() {
        try {
            const element = await this.driver.wait(until.elementLocated(this.productLayout), 10000);
            return await element.isDisplayed();
        } catch (error) {
            return false;
        }
    }

    /**
     * Check if no results message is visible
     */
    async isNoResultsMessageVisible() {
        try {
            const element = await this.driver.wait(until.elementLocated(this.noResultsMessage), 10000);
            return await element.isDisplayed();
        } catch (error) {
            return false;
        }
    }

    /**
     * Check if product thumbs are not present
     */
    async areProductThumbsNotPresent() {
        try {
            const elements = await this.driver.findElements(this.productThumb);
            return elements.length === 0;
        } catch (error) {
            return true;
        }
    }

    /**
     * Check if showing results text is visible
     */
    async isShowingResultsTextVisible() {
        try {
            const element = await this.driver.wait(until.elementLocated(this.showingResultsText), 10000);
            return await element.isDisplayed();
        } catch (error) {
            return false;
        }
    }

    /**
     * Get showing results text
     */
    async getShowingResultsText() {
        try {
            const element = await this.driver.wait(until.elementLocated(this.showingResultsText), 10000);
            return await element.getText();
        } catch (error) {
            return '';
        }
    }

    /**
     * Get current URL
     */
    async getCurrentUrl() {
        return await this.driver.getCurrentUrl();
    }
}

module.exports = EcommercePage;

File: ui-automation/selenium-java/src/main/java/pages/EcommercePage.java
package pages;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.openqa.selenium.support.ui.ExpectedConditions;
import utils.DriverFactory;

/**
 * Page Object Model for eCommerce Playground Homepage
 */
public class EcommercePage {
    private WebDriver driver;
    private WebDriverWait wait;
    
    // Page elements
    @FindBy(name = "search")
    private WebElement searchInput;
    
    @FindBy(xpath = "//span[@class='title' and contains(text(), ' Home')]")
    private WebElement homeButton;
    
    @FindBy(xpath = "//li[contains(text(), ' Special')]")
    private WebElement specialButton;
    
    @FindBy(xpath = "//li[contains(text(), ' Blog')]")
    private WebElement blogButton;
    
    @FindBy(xpath = "//li[contains(text(), ' Mega Menu')]")
    private WebElement megaMenuButton;
    
    @FindBy(xpath = "//h1[@class='h3' and contains(text(), 'Special Offers')]")
    private WebElement specialOffersHeader;
    
    @FindBy(xpath = "//h3[contains(text(), 'Latest Articles')]")
    private WebElement latestArticlesHeader;
    
    @FindBy(xpath = "//img[@alt='Sony VAIO']")
    private WebElement sonyVaioImage;
    
    @FindBy(className = "product-thumb")
    private WebElement productThumb;
    
    @FindBy(className = "product-layout")
    private WebElement productLayout;
    
    @FindBy(xpath = "//div[contains(text(), 'There is no product that matches the search criteria.')]")
    private WebElement noResultsMessage;
    
    @FindBy(xpath = "//div[@class='text-right' and contains(text(), 'Showing')]")
    private WebElement showingResultsText;
    
    // Constructor
    public EcommercePage() {
        this.driver = DriverFactory.getDriver();
        this.wait = DriverFactory.getWait();
        PageFactory.initElements(driver, this);
    }
    
    // Navigation methods
    public void navigateToHomepage() {
        driver.get("https://ecommerce-playground.lambdatest.io/");
    }
    
    public void clickHomeButton() {
        wait.until(ExpectedConditions.elementToBeClickable(homeButton)).click();
    }
    
    public void clickSpecialButton() {
        WebElement specialElement = wait.until(ExpectedConditions.elementToBeClickable(
            By.xpath("//div[@id='widget-navbar-217834']//li[contains(text(), ' Special')]")));
        specialElement.click();
    }
    
    public void clickBlogButton() {
        WebElement blogElement = wait.until(ExpectedConditions.elementToBeClickable(
            By.xpath("//div[@id='widget-navbar-217834']//li[contains(text(), ' Blog')]")));
        blogElement.click();
    }
    
    public void clickMegaMenuButton() {
        WebElement megaMenuElement = wait.until(ExpectedConditions.elementToBeClickable(
            By.xpath("//div[@id='widget-navbar-217834']//li[contains(text(), ' Mega Menu')]")));
        megaMenuElement.click();
    }
    
    // Search functionality
    public void performSearch(String searchTerm) {
        wait.until(ExpectedConditions.elementToBeClickable(searchInput)).click();
        searchInput.clear();
        searchInput.sendKeys(searchTerm);
        searchInput.submit();
    }
    
    // Verification methods
    public boolean isSearchInputVisible() {
        return wait.until(ExpectedConditions.visibilityOf(searchInput)).isDisplayed();
    }
    
    public boolean isHomeButtonVisible() {
        return homeButton.isDisplayed();
    }
    
    public boolean isSpecialOffersHeaderVisible() {
        return wait.until(ExpectedConditions.visibilityOf(specialOffersHeader)).isDisplayed();
    }
    
    public boolean isLatestArticlesHeaderVisible() {
        return wait.until(ExpectedConditions.visibilityOf(latestArticlesHeader)).isDisplayed();
    }
    
    public boolean isSonyVaioImageVisible() {
        return wait.until(ExpectedConditions.visibilityOf(sonyVaioImage)).isDisplayed();
    }
    
    public boolean areProductsVisible() {
        return wait.until(ExpectedConditions.visibilityOf(productLayout)).isDisplayed();
    }
    
    public boolean isNoResultsMessageVisible() {
        return wait.until(ExpectedConditions.visibilityOf(noResultsMessage)).isDisplayed();
    }
    
    public boolean areProductThumbsNotPresent() {
        try {
            return driver.findElements(By.className("product-thumb")).isEmpty();
        } catch (Exception e) {
            return true;
        }
    }
    
    public boolean isShowingResultsTextVisible() {
        return wait.until(ExpectedConditions.visibilityOf(showingResultsText)).isDisplayed();
    }
    
    public String getShowingResultsText() {
        return showingResultsText.getText();
    }
    
    public String getCurrentUrl() {
        return driver.getCurrentUrl();
    }
}
